### ISSUE TITLE
MsQuicStream can leak CancellationTokenRegistrations

### ISSUE THREAD

**Issue by MihaZupan** [dotnet/runtime#69491]
While reading through #69483 I noticed that [SetupWriteStartState](https://github.com/dotnet/runtime/blob/a598ebb67973f03b1d6b3fd6c50f75bae3fe1c51/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L313) creates a `CancellationTokenRegistration`, but the method [may still throw](https://github.com/dotnet/runtime/blob/a598ebb67973f03b1d6b3fd6c50f75bae3fe1c51/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L366-L389), in which case the registration won't be disposed.

As a result the `MsQuicStream.State` will remain rooted by the (potentially user-provided) cancellation token.

cc: @rzikm @ManickaP

**Comment by rzikm** [comment]
Duplicate of https://github.com/dotnet/runtime/issues/55437

I want to tackle this once https://github.com/dotnet/runtime/pull/68772 is merged, we can do additional refactorings once we don't have 3 duplicate Send*Async implementations
